### ISSUE TITLE
Skip symbolic links when passing files to Prettier

### ISSUE THREAD

**Issue by yermulnik** [super-linter/super-linter#6378]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

![image](https://github.com/user-attachments/assets/4dbfdf8d-22fb-4489-aa02-731bbdf3a3db)

Repro:
```shell
> file test.yaml
test.yaml: symbolic link to examples/account-request-template.yaml

> prettier -c ./test.yaml ; echo $?
Checking formatting...
[error] Explicitly specified pattern "./test.yaml" is a symbolic link.
All matched files use Prettier code style!
2
```

### Expected Behavior

`super-linter` should bypass `prettier` failure to scan symbolic links.

### Anything else?

There are two options that I envision:
* Add `--no-error-on-unmatched-pattern` cmdline flag to `PRETTIER_COMMAND=(prettier)` at https://github.com/super-linter/super-linter/blob/main/lib/functions/linterCommands.sh#L16
* Add test for input file to be plain file (or to be not a symlink) where `PRETTIER_COMMAND` is run for each particular type of input file

Option 1 looks simple, though may suppress something else (Prettier's docu is a bit vague saying just «_Prevent errors when pattern is unmatched_»: https://github.com/prettier/prettier/blob/main/docs/cli.md#user-content---no-error-on-unmatched-pattern)
The 2nd option looks a bit more robust, though would require change to the way files are passed to Prettier and may result in a more complex code change.

Please let me know which option looks best for you and I'll try and create PR for that. Thanks.

Refs:
* https://github.com/prettier/prettier/issues/15723
* https://github.com/prettier/prettier/issues/15843
* https://github.com/prettier/prettier/pull/15533
* https://prettier.io/docs/en/cli.html#file-patterns («_Prettier CLI will not follow symbolic links when expanding arguments_»)
* https://github.com/prettier/prettier/blob/main/docs/cli.md#user-content---no-error-on-unmatched-pattern
* https://github.com/prettier/prettier/blob/main/CHANGELOG.md#skip-explicitly-passed-symbolic-links-with---no-error-on-unmatched-pattern-15533-by-sanmai-nl

**Comment by ferrarimarco** [comment]
Thanks for raising this. I experienced this exact issue in other repositories, too.

The workaround would be to use a [`.prettierignore` file](https://prettier.io/docs/en/ignore.html#ignoring-files-prettierignore) to ignore symbolic links.

[This Prettier PR](https://github.com/prettier/prettier/pull/14627) contains some interesting insights about why Prettier doesn't follow symlinks. Some of them also apply to super-linter, so your second option might make sense.

We might also think about applying in general, and not only to Prettier because other linters have issues with symlinks as well. Maybe this would be considered a breaking change, tho?

@zkoppert @Hanse00 WDYT?

**Comment by yermulnik** [comment]
> The workaround would be to use a [`.prettierignore` file](https://prettier.io/docs/en/ignore.html#ignoring-files-prettierignore) to ignore symbolic links.

That isn't an option (or I'm doing something wrong):
```shell
>  file test.yaml
test.yaml: symbolic link to examples/account-request-template.yaml

> cat .prettierignore
test.yaml

> prettier -c ./test.yaml
Checking formatting...
[error] Explicitly specified pattern "./test.yaml" is a symbolic link.
All matched files use Prettier code style!

> prettier --ignore-path ./.prettierignore -c test.yaml
Checking formatting...
[error] Explicitly specified pattern "test.yaml" is a symbolic link.
All matched files use Prettier code style!
```

**Comment by ferrarimarco** [comment]
Can you try using the `FILTER_REGEX_EXCLUDE` variable (it's a super-linter config variable) to ignore the symlink? This might unblock you as we think about this.

**Comment by yermulnik** [comment]
@ferrarimarco Thanks for the pointer. I'll look into this if you'll decide to postpone resolution to later.

**Comment by yermulnik** [comment]
> as we think about this.

Hi @ferrarimarco. Have there been any updates or outcome re this? Thanks.

**Comment by zkoppert** [comment]
Yeah I would consider this a breaking change and agree with the approach described.

**Comment by ferrarimarco** [comment]
I think we don't want to exclude symbolic links for every file. Considering that there's already a work around for this issue (using `.prettierignore`), we can close this one.

**Comment by yermulnik** [comment]
@ferrarimarco Not really sure which workaround using `.prettierignore` you refer to. I guess this https://github.com/super-linter/super-linter/issues/6378#issuecomment-2489174988 comment of mine showcases that the workaround of adding symlinks to `.prettierignore` doesn't work and `prettier` throws error either way 🤷🏻 

The other workaround using `FILTER_REGEX_EXCLUDE` looks viable, though it brings a bunch of toil if using `super-linter` workflow from another repository (like e.g. at my current job we maintain `super-linter` workflow in a central repo and apply it via Org-level repository ruleset to all DevOps repos to avoid maintaining `super-linter` per each repo individually). While additionally `FILTER_REGEX_EXCLUDE` is a way too global exclude rather than prettier-specific.

FWIW: adding `--no-error-on-unmatched-pattern` to `prettier`'s cmdline args should be the easiest solution if I read it correct: https://github.com/prettier/prettier/pull/15533/files

**Comment by ferrarimarco** [comment]
I meant using `FILTER_REGEX_EXCLUDE`, not `.prettierignore`. I updated my comment above to reflect that.

One way to use `FILTER_REGEX_EXCLUDE` with a "global" super-linter job might be to add an input variable to pass a specific `FILTER_REGEX_EXCLUDE` for the repositories that need that.

Let me reopen this one while we think a bit more.

**Comment by ferrarimarco** [comment]
I think we can add an option to avoid following symbolic links, disabled by default.

**Comment by yermulnik** [comment]
Just a tiny note: it is not about following symlinks, it is about not feeding them to Prettier.
So when we add a symlink to the repo, there's no point in linting symlink target if it hasn't been changed (which, in case it has been changed, is being fed to linter as a standalone entity).

What I mean is that this issue is exactly about skipping symlinks for Prettier (and possibly for other linters that have the same odd behavior around symlinks, but that I don't know of). Somewhat this pseudo-code:
```shell
if [ -L "$FILE" -a "${LINTER[0}" = "prettier" ]; then
    debug "Skip \"$FILE\" when running \"${LINTER[0]}\""
else
    <run linter as usual>
fi
```

Along with that an other option to specifically avoid **following** symlinks globally might also make sense but is of a different essence.

**Comment by yermulnik** [comment]
If the patch above looks reasonable, I'll create a PR for that and will appreciate some general guidance on how to test the change locally as I haven't found any guidelines or refs to guidelines in README. Thank you.

**Comment by claudioandre-br** [comment]
My two cents.

I'm using FILTER_REGEX_EXCLUDE (for each repository ...) and **it works perfectly well**.

But why should super-linter pass a file to prettier if we all know it shouldn't?

**Comment by yermulnik** [comment]
> But why should super-linter pass a file to prettier if we all know it shouldn't?

💯

**Comment by ferrarimarco** [comment]
The idea is ok, although the patch you proposed will not work because `FILE_TYPE` is the file extension, not the linter/formatter.

I'd suggest:

Create a function (say `IsSupportedByPrettier` or something like that) in `detectFiles.sh` to check if a file is something that Prettier supports. Add tests for this function in `test/lib/detectFilesTest.sh`. This function should test if the file has one of the extensions that Prettier supports, and if it's not a symlink

Then, centralize the `echo "${FILE}" >>"${FILE_ARRAYS_DIRECTORY_PATH}/file-array-XXXX_PRETTIER"` statements in `buildFileList.sh` under an if statement:

    ```shell
    if IsSupportedByPrettier "${FILE}"; then
      echo "${FILE}" >>"${FILE_ARRAYS_DIRECTORY_PATH}/file-array-HTML_PRETTIER"
      echo "${FILE}" >>"${FILE_ARRAYS_DIRECTORY_PATH}/file-array-JAVASCRIPT_PRETTIER"
      ....
    fi
    ```

Hope this helps!

**Comment by yermulnik** [comment]
@ferrarimarco Appreciate the pointers. They are really useful.

I've gone with similar but a bit different imlementation: https://github.com/super-linter/super-linter/pull/6620 — WDYT? 🤔
